pkg/imgutil/dockerconfigresolver/credentialsstore.go

```diff
@@ -0,0 +1,182 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package dockerconfigresolver
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/cli/cli/config/types"
+
+	"github.com/containerd/log"
+)
+
+type Credentials = types.AuthConfig
+
+// NewCredentialsStore returns a CredentialsStore from a directory
+// If path is left empty, the default docker `~/.docker/config.json` will be used
+// In case the docker call fails, we wrap the error with ErrUnableToInstantiate
+func NewCredentialsStore(path string) (*CredentialsStore, error) {
+	dockerConfigFile, err := config.Load(path)
+	if err != nil {
+		return nil, errors.Join(ErrUnableToInstantiate, err)
+	}
+
+	return &CredentialsStore{
+		dockerConfigFile: dockerConfigFile,
+	}, nil
+}
+
+// CredentialsStore is an abstraction in front of docker config API manipulation
+// exposing just the limited functions we need and hiding away url normalization / identifiers magic, and handling of
+// backward compatibility
+type CredentialsStore struct {
+	dockerConfigFile *configfile.ConfigFile
+}
+
+// Erase will remove any and all stored credentials for that registry namespace (including all legacy variants)
+// If we do not find at least ONE variant matching the namespace, this will error with ErrUnableToErase
+func (cs *CredentialsStore) Erase(registryURL *RegistryURL) (map[string]error, error) {
+	// Get all associated identifiers for that registry including legacy ones and variants
+	logoutList := registryURL.AllIdentifiers()
+
+	// Iterate through and delete them one by one
+	errs := make(map[string]error)
+	for _, serverAddress := range logoutList {
+		if err := cs.dockerConfigFile.GetCredentialsStore(serverAddress).Erase(serverAddress); err != nil {
+			errs[serverAddress] = err
+		}
+	}
+
+	// If we succeeded removing at least one, it is a success.
```

Comment:
what's the current behavior if the removal partially succeeds (some are removed and some causes errors)?

Reply:
Also if some are left un-removed. We should at least log those failed cases for awareness.

Reply:
I think it's okay as long as it doesn't regress existing behaviors

Reply:
Agreed. I do trust our current testing (to some extent). 
I will make sure to further extend testing for these things when I will move to part 2 (some work in progress in there: - not ready for review at all, but that gives an idea -> #3249)